File: 376. Wiggle Subsequence.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var wiggleMaxLength = function(nums) {
    if (nums.length === 0) return 0;
    let T = [], diff = [];
    let res = 1;
    for (let i = 0; i < nums.length; i++)
    {
        T.push(1);
        diff.push(0);
    }
    for (let i = 1; i < nums.length; i++)
    {
        for (let j = 0; j < i; j++)
        {
            if (diff[j] === 0 || (diff[j] > 0 && nums[i] < nums[j]) || (diff[j] < 0 && nums[i] > nums[j]))
            {
                T[i] = Math.max(T[i], T[j] + 1);
                diff[i] = nums[i] - nums[j];
            }
        }
        res = Math.max(res, T[i]);
    }
    return res;
};

File: 376. Wiggle Subsequence.py
class Solution(object):
    def wiggleMaxLength(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if not nums:
            return 0
        T = [1 for i in xrange(len(nums))]
        diff = [0 for i in xrange(len(nums))]
        res = 1
        for i in xrange(1, len(nums)):
            for j in xrange(i):
                if nums[i] != nums[j]:
                    if diff[j] == 0:
                        T[i] = max(T[i], T[j] + 1)
                        if nums[i] > nums[j]:
                            diff[i] = 1
                        else:
                            diff[i] = -1
                    elif diff[j] == 1 and nums[i] < nums[j]:
                        T[i] = max(T[i], T[j] + 1)
                        diff[i] = -1
                    elif diff[j] == -1 and nums[i] > nums[j]:
                        T[i] = max(T[i], T[j] + 1)
                        diff[i] = 1
            res = max(res, T[i])
        return res
